File: src/main/java/com/github/tototoshi/csv/SourceLineReader.java
package com.github.tototoshi.csv;

import scala.collection.Iterator;
import scala.io.Source;

import java.io.IOException;

/**
 * Splits lines over an internally owned buffer.
 *
 * Source offers no bulk read, so the refill still pulls one char at a time;
 * the win is that it pulls them into a char[] rather than appending each to a
 * StringBuilder, and that it goes through Source.iter() rather than
 * Source.next(). The latter matters more than it looks: Source.next() runs a
 * Positioner per character, encoding a position and tracking line and column,
 * all of which a line reader throws away.
 */
public class SourceLineReader extends AbstractBufferedLineReader {

    private final Source source;
    private final Iterator<Object> iter;

    public SourceLineReader(Source source) {
        this(source, DEFAULT_BUFFER_SIZE);
    }

    SourceLineReader(Source source, int bufferSize) {
        super(bufferSize);
        this.source = source;
        this.iter = source.iter();
    }

    @Override
    protected int fillBuffer(char[] dst) {
        int n = 0;
        while (n < dst.length && iter.hasNext()) {
            dst[n++] = (Character) iter.next();
        }
        return n == 0 ? -1 : n;
    }

    @Override
    public void close() throws IOException {
        source.close();
    }
}

File: src/main/scala/com/github/tototoshi/csv/SourceLineReader.scala
package com.github.tototoshi.csv

import scala.io.Source

class SourceLineReader(source: Source) extends LineReader {

  private var currentChar: Int = -1

  override def readLineWithTerminator(): String = {
    val sb = new StringBuilder
    while ({
      if (!source.hasNext && currentChar == -1) {
        if (sb.isEmpty) return null
        else return sb.toString
      }
      val c: Int =
        if (currentChar > -1) {
          val ch = currentChar
          currentChar = -1
          ch
        } else {
          source.next().toInt
        }

      sb.append(c.toChar)

      if (c == '\n' || c == '\u2028' || c == '\u2029' || c == '\u0085') {
        return sb.toString
      }

      if (c == '\r') {
        if (!source.hasNext) return sb.toString
        val next = source.next()
        if (next != '\n') currentChar = next
        return sb.toString
      }

      true
    }) {}
    sb.toString
  }

  override def close(): Unit = source.close()
}
